docs/architecture/adr-100-data-companion-push-api.md

```diff
@@ -0,0 +1,441 @@
+    optional tendermint.abci.FinalizeBlockResponse finalize_block_response = 2;
+}
+
+// BlockCommittedResponse is either empty upon succes, or returns one or more
+// errors. Note that returning any errors here will cause Tendermint to crash.
+message BlockCommittedResponse {
+    // Zero or more errors occurred during the companion's processing of the
+    // request.
+    repeated Error errors = 1;
+}
+
+// Error can be one of several different types of errors.
+message Error {
+    oneof error {
+        UnexpectedBlockError unexpected_block_err = 1;
+        UnexpectedFieldError unexpected_field_err = 2;
+        ExpectedFieldError   expected_field_err   = 3;
+    }
+}
+
+// UnexpectedBlockError is returned by the server when Tendermint sent it a
+// block that is ahead of the block expected by the server.
+message UnexpectedBlockError {
+    // The height of the block expected by the server.
+    int64 expected_height = 1;
+}
+
+// UnexpectedFieldError is returned by the server when Tendermint sent it a
+// message containing an unexpected field. For instance, if the companion
+// expects only the `block` field but not `finalize_block_response`, but somehow
+// the companion receives a value for `finalize_block_response`, it should
+// return this error because it indicates that either Tendermint or the data
+// companion are most likely incorrectly configured.
+message UnexpectedFieldError {
+    // The name of the field whose value was expected to be empty, but was not.
+    string field_name = 1;
+}
+
+// ExpectedFieldError is returned by the server when Tendermint sent it a
+// message containing an empty field value for a field it was expecting to be
+// populated. If this occurs, it indicates that either Tendermint or the data
+// companion are most likely incorrectly configured.
+message ExpectedFieldError {
+    // The name of the field whose value was expected to be populated, but was
+    // empty.
+    string field_name = 1;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should facilitate buffering of data destined for the companion
+service. This data would need to be stored on disk to ensure that Tendermint
+could attempt to resubmit all unsubmitted data upon Tendermint startup.
+
+At present, this data is already stored on disk, and so practically we would
+need to implement some form of background pruning mechanism to remove the data
+we know has been shared with the companion service.
+
+In order to not allow disk storage to grow unbounded, _and_ ensure that we do
+not lose any critical data, we need to impose some kind of limit on the number
+of requests we will buffer before crashing the Tendermint node. In case there is
+manual intervention needed to restore the companion service, Tendermint should
+attempt to flush this buffer upon startup before continuing normal operations.
+
```

Comment:
Also, it would be wise IMO to define the use case where the companion is started for the first time, much _after_ CometBFT itself was started. Similarly to how state-sync works today, the companion should be aware that, in such a use case, it would have a "gap" in its history.
Later version may improve on this, but this simplistic use case should is as good as what we have today when state-sync is used.